test/test_sum_product.py

```diff
@@ -86,21 +87,616 @@ def test_partition(inputs, dims, expected_num_components):
-    factors2 = partial_sum_product(sum_op, prod_op, factors1, vars2, plates)
+    factors1 = impl(sum_op, prod_op, factors, vars1, plates)
+    factors2 = impl(sum_op, prod_op, factors1, vars2, plates)
     actual = reduce(prod_op, factors2)
 
     expected = sum_product(sum_op, prod_op, factors, vars1 | vars2, plates)
     assert_close(actual, expected)
 
 
+def _expected_hmm_example(sum_op, prod_op, factors, plates, global_vars, local_vars,
+                          time, duration, step):
+
+    step = OrderedDict(sorted(step.items()))
+    drop = tuple("_drop_{}".format(i) for i in range(len(step)))
+    prev_to_drop = dict(zip(step.keys(), drop))
+    curr_to_drop = dict(zip(step.values(), drop))
+
+    unrolled_factors = []
+    reduce_vars = global_vars | plates
+    reduce_vars |= frozenset(
+        ('{}_{}'.format(var, t+1) for
+         var in local_vars for t in range(duration))
+    )
+    reduce_vars |= frozenset(
+        ('{}_{}'.format(curr_to_drop[var], t+1) for
+         var in curr_to_drop.keys() for t in range(duration))
+    )
+    reduce_vars |= frozenset(
+        ('{}_{}'.format(prev_to_drop[var], t) for
+         var in prev_to_drop.keys() for t in range(duration))
+    )
+    for factor in factors:
+        if time in factor.inputs:
+            slice_factors = [factor(
+                **{time: t},
+                **{var: '{}_{}'.format(var, t+1) for var in local_vars},
+                **{var: '{}_{}'.format(curr_to_drop[var], t+1) for var in curr_to_drop.keys()},
+                **{var: '{}_{}'.format(prev_to_drop[var], t) for var in prev_to_drop.keys()}
+                ) for t in range(factor.inputs[time].size)]
+            unrolled_factors.extend(slice_factors)
+        else:
+            unrolled_factors.append(factor)
+    expected = sum_product(sum_op, prod_op, unrolled_factors, reduce_vars, plates)
+
+    return expected
+
+
+@pytest.mark.parametrize('vars1,vars2', [
+    (frozenset(),
+     frozenset({"sequences", "time", "x_prev", "x_curr", "tones", "y_prev", "y_curr"})),
+    (frozenset({"time", "x_prev", "x_curr", "tones", "y_prev", "y_curr"}),
+     frozenset({"sequences"})),
+    (frozenset({"sequences", "time", "x_prev", "x_curr", "y_prev", "y_curr"}),
+     frozenset()),
+])
+@pytest.mark.parametrize('sum_op,prod_op', [(ops.logaddexp, ops.add), (ops.add, ops.mul)])
+def test_partial_sum_product_hmm_example_0(sum_op, prod_op, vars1, vars2):
+    x_dim, y_dim = 2, 3
+    sequences, duration, tones = 2, 5, 4
+
+    probs_x = random_tensor(OrderedDict({}))
+    probs_y = random_tensor(OrderedDict({}))
+
+    x = random_tensor(OrderedDict({
+        "sequences": Bint[sequences],
+        "time": Bint[duration],
+        "x_prev": Bint[x_dim],
+        "x_curr": Bint[x_dim],
+    }))
+
+    y = random_tensor(OrderedDict({
+        "sequences": Bint[sequences],
+        "time": Bint[duration],
+        "tones": Bint[tones],
+        "y_prev": Bint[y_dim],
```

Comment:
Note this `y_curr` and the `sequences` plate means this test does not correspond to model 0, although I think it's still a reasonable test case since factors `x` and `y` are completely disconnected.